### PR TITLE
Cleanup project tags

### PR DESCRIPTION
**Pull request by davidfischer**
- Adds a management command to clean up tags. Cleanup involves lowercasing, slugifying (canonicalizing) and removing any tags with no project (no tagged items, specifically)
- Future tags will come in canonicalized
- Adds a management command to get tags from github for projects where we have no tags.
- Adds an admin action to import tags from github that shares logic with above.

## Notes & Possible Enhancements

* GitHub actually limits the length of the tag which might not be a bad idea. Basically all tags longer than ~30 characters are certainly spam.

    Screen Shot 2019-07-23 at 11 01 10 AM

* Limiting the number of tags for a project to some reasonable number like 20 isn't a bad idea.
* I would definitely like to autocomplete tags but I think that's a task best left until after other design work

### CODE REVIEW

**Comment by agjohnson** [comment]
Diving into review now, but immediately:

* Is there any reason not to make commands that will run a single time a data migration instead? If new tags are formatted correctly, we'll need to manually run this command once. If we make this a migration, it happens for all instances automatically. Would this be bad?
* Is there any reason not to make recurring management commands an admin function instead?

Generally speaking, I think site admin is a better place for most management commands, especially those that don't require input. These commands are easier to find and keeps site admin role abilities encapsulated in the site admin, instead of requiring ops access to servers for some selective functions. These functions are generally lost.

If that is not the correct answer here, that is fine. But I like to steer us clear of management commands where possible. All of our management commands suffer from rot and usually reimplement logic that admin functions give us.

**Comment by davidfischer** [comment]
> Is there any reason not to make recurring management commands an admin function instead?

Cleaning tags will probably take on the order of 1-2 hours. Importing tags from github will take on the order of 1-3 days at least the first time. Incremental may be only 1 dayish. I'm not sure an admin function or a data migration is appropriate for that. We could implement it as an admin function that calls the same logic on a per project basis but I do want to run this for all projects.

> If new tags are formatted correctly, we'll need to manually run this command once.

This is true for the cleaning aspect but not for deleting unused tags. Deleting unused tags should be fairly quick incrementally. When tags are first cleaned, there will be a lot to delete. For example, all tags with uppercase will be deleted. However, when a bunch of spam projects get deleted in the future, they will leave behind a bunch of spam tags which this management command will clean up on future runs. Possibly reformatting tags and removing unused ones should be separated.

**Comment by agjohnson** [comment]
Roger, thanks for the background :+1: 

I think my reservations are mostly around frequency, but the length of run is another point.

Even a 1-2 hour migration is pretty annoying, so a management command is ok, though it seems redundant functionality.

For the tag import, what is the frequency we want to run this? Just once for now to test this data? A management command works as a first pass for bulk import, but I'd want to move this out of a management command eventually if we're running this frequently or this needs to be a scheduled task that ops needs to take. If we're working towards running this more frequently, there are plenty of options to automate this via celery.

I'm guessing this would be something we're running once for now, would it be helpful to run this selectively on projects in addition? This also sounds like later work.

**Comment by davidfischer** [comment]
> I'm guessing this would be something we're running once for now, would it be helpful to run this selectively on projects in addition? This also sounds like later work.

For the import from github, I'm looking mostly just to run this once although once a year doesn't seem too bad. Having the ability to do it on a per project basis would be nice though. At the same time, connecting to the github API requires an API token which is separate from an app token (I think). Otherwise, the [rate limit is ~60/hr](https://developer.github.com/v3/#rate-limiting).

Canonicalizing tags only ever needs to run once after this is merged as future tags will be canonicalized. Removing unused tags is something I would probably like to run every time we delete a bunch of spam projects. Once spam deletion is more automated, this could also be automated. Deleting unused tags should also be pretty fast overall (a few minutes maybe, I'm guesstimating).

To make this more concrete, how about this proposal:

- Leave the import from github as a management command. This takes a long time and is run infrequently. It also requires a github API token. If in the future we have this token, we could add the ability to run this on a per project basis from the admin.
- Add an admin function to delete unused tags in core.
- Leave the canonicalizing tags as a management command. It only needs to be run once. At the end, it will call the same logic as the admin function to delete unused tags.

**Comment by davidfischer** [comment]
> At the same time, connecting to the github API requires an API token which is separate from an app token (I think). 

It looks like [we can use our app credentials](https://developer.github.com/v3/#increasing-the-unauthenticated-rate-limit-for-oauth-applications) (client ID and client secret). It doesn't give any permissions but does increase the rate limit which is all we need for this.

**Comment by davidfischer** [comment]
> Add an admin function to delete unused tags in the projects app.

This is a bit awkward as you **must** select projects in order to do it and then those are ignored. Maybe this is best as a management command for now until spam deletion is automated and then this can be a celery task.

Screen Shot 2019-07-25 at 11 34 43 AM

**Comment by agjohnson** [comment]
> This is a bit awkward as you must select projects in order to do it and then those are ignored

Yeah I agree, it sounds like management command works fine here. There's probably a pattern for admin action against *all* projects, short of selecting "all projects" in the UI selector. I think ultimately full automation would the way to go with some of these functoins, and leave fiddling with the admin out of it. I do see some potential to selectively choose Projects from the admin though.

**Comment by davidfischer** [comment]
> I do see some potential to selectively choose Projects from the admin though.

I added an admin function to get tags for a single project and used the same logic in the management command.

Regardless, I think this is ready for a full review.

File: readthedocs/core/management/commands/import_github_tags.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+from readthedocs.projects.models import Project
+from readthedocs.projects.tag_utils import import_tags
+
+
+class Command(BaseCommand):
+
+    help = __doc__
+
+    def handle(self, *args, **options):
+        queryset = Project.objects.filter(tags=None).filter(repo__contains='github.com')
```
[the remark this answers is not in the document]
Comment (davidfischer) [reply readthedocs/core/management/commands/import_github_tags.py]: I tend to agree and I wouldn't mind expanding them to the other providers.

**Pull request merged**